lib/Dialect/Linalg/Transforms/Tiling.cpp

```diff
@@ -215,9 +215,28 @@ makeTiledViews(OpBuilder &b, Location loc, LinalgOp linalgOp,
   return res;
 }
 
+bool isPermutationValid(const ArrayRef<uint64_t>& permutation) {
+  SmallVector<bool, 4> present(permutation.size(), false);
+  for (unsigned i = 0; i < permutation.size(); ++i) {
+    if (permutation[i] >= permutation.size())
+      return false;
+    present[i] = true;
+  }
+  return std::all_of(present.begin(), present.end(), [](bool v) { return v; });
```

Comment:
What is this doing? When are any of these ever going to be false?

Reply:
Check is required in case of duplicated entries in the permutation array (which is not a valid permutation). Added comments to clarify both here and in Utils.h.
Would it be better to write the loop instead of using std:all_of?
Also, is there a way to enforce these restrictions at compile time in TableGen?

Reply:
Did you mean to use present[permutation[i]]?

Also if you don't use SmallBitVector like suggested above, the std::all_of should be changed to either llvm::all_of(...) or !llvm::is_contained(present, false);

Reply:
This seems to point to a lack of proper test coverage.

Reply:
ok, I refrained for suggesting this earlier but given @River707's comment can we just drop this function, turn into an AffineMap (add a helper in AffineMap.cpp if necessary) and just use `isProjectedPermutationMap`?
See: https://github.com/tensorflow/mlir/blob/master/lib/IR/AffineMap.cpp#L244

Reply:
The explicit for loop  was checking for out of range values in `permutation`, while marking the presence of valid values. But, if duplicates were present, some `present` elements could remain `false` after the loop.
For example, if `permutation=[0,3,1]`, the `if`condition in the  `for` loop will prevent an invalid access (since `3 > permutation.size()` ) and return false. But, if  `permutation=[0,1,1]`, then `present[2]==false` after the for loop, so std::all_of will return false.

Regarding the use of `isProjectedPermutationMap` (or `isPermutation`), there is no explicit check for out-of-bound accesses and it lets `SmallVector` fail, obscuring the reason for failure. I.e. if `permutation=[0,3,1]`, the AffineMap is successfully created, but  checking `if (seen[dim.getPosition()])` in `isProjectedPermutationMap` will access out-of-bounds triggering an `assert` in SmallVector.h.

I am happy either way, but I thought I should make you aware of this. Is it worth adding the extra guard to `isProjectedPermutationMap`?

Reply:
There is a bug in the creation of the AffineMap:  you should not do 
`AffineMap::get(permutation.size(), 0, perm)` but
`AffineMap::get(getMax(permutation), 0, perm)`

The AffineMap you describe should be:
`(d0,  d1, d2) -> (d0, d3, d1)` which is an invalid map (`d3` is an undefined coordinate dimension).

If this is the case, the potential fact that it is not caught by some verifier at the type level should be fixed.

In the meantime, this seems like an easy enough pitfall to fall into (and to miss when reviewing): let's factor out this code and have a permutation builder helper in AffineMap.h that would build a valid map from a permutation array.

Thanks!

Reply:
You are right, thank you. We have implemented a builder helper in AffineMap.h.